packages/loopring_v3/contracts/amm/libamm/AmmExitProcess.sol

```diff
@@ -0,0 +1,169 @@
+        )
+        internal
+    {
+        // Check that the withdrawal in the block matches the expected withdrawal
+        WithdrawTransaction.Withdrawal memory withdrawal = ctx._block.readWithdrawal(ctx.txIdx++);
+        ctx.numTransactionsConsumed++;
+
+        // These fields are not read by readWithdrawal: storageID
+        withdrawal.minGas = 0;
+        withdrawal.to = address(this);
+        withdrawal.extraData = new bytes(0);
+
+        require(
+            withdrawal.withdrawalType == 1 &&
+            withdrawal.owner == address(this) &&
+            withdrawal.accountID == S.accountID &&
+            withdrawal.tokenID == token.tokenID &&
+            withdrawal.amount == amount && //No rounding errors because we put in the complete uint96 in the DA.
+            withdrawal.feeTokenID == 0 &&
+            withdrawal.fee == 0 &&
+            withdrawal.onchainDataHash == WithdrawTransaction.hashOnchainData(
+                withdrawal.minGas,
+                withdrawal.to,
+                withdrawal.extraData
+            ),
+            "INVALID_TX_DATA"
+        );
+
+        // Now approve this withdrawal
+        withdrawal.validUntil = 0xffffffff;
+        bytes32 txHash = WithdrawTransaction.hashTx(ctx.exchangeDomainSeparator, withdrawal);
+        ctx.exchange.approveTransaction(address(this), txHash);
+
+        // Total balance in this contract increases by the amount withdrawn
+        S.totalLockedBalance[token.addr] = S.totalLockedBalance[token.addr].add(amount);
+    }
+
+    function processExit(
+        AmmData.State    storage S,
+        AmmData.Context  memory  ctx,
+        AmmData.PoolExit memory  exit,
+        bytes            memory  signature
+        )
+        internal
+    {
+        S.validatePoolTransaction(
+            exit.owner,
+            AmmExitRequest.hashPoolExit(ctx.domainSeparator, exit),
+            signature
+        );
+
+        if (signature.length == 0) {
+            // This is an onchain exit, we're processing it now so stop tracking it.
+            S.isExiting[exit.owner] = false;
+        }
+
+        (bool slippageRequirementMet, uint96[] memory amounts) = _validateExitAmounts(S, ctx, exit);
+
+        if (!slippageRequirementMet) return;
+
+        for (uint i = 0; i < ctx.size; i++) {
+            uint96 amount = amounts[i];
+            if (exit.toLayer2) {
+                TransferTransaction.Transfer memory transfer = ctx._block.readTransfer(ctx.txIdx++);
+                ctx.numTransactionsConsumed++;
+
+                require(
+                    transfer.fromAccountID == S.accountID &&
+                    transfer.from == address(this) &&
+                    transfer.to == exit.owner &&
+                    transfer.tokenID == ctx.tokens[i].tokenID &&
+                    transfer.amount.isAlmostEqual(amount) &&
+                    transfer.feeTokenID == ctx.tokens[i].tokenID &&
+                    transfer.fee.isAlmostEqual(exit.fee),
```

Comment:
For exits the fee should be subtracted from the amount withdrawn for the user, otherwise it's the pool account that pays the fee: `amount := transfer.fee + transfer.amount`.

Reply:
removed exit fees